Sales Invoice | test_pulling_advance_based_on_debit_to | frappe.exceptions.ValidationError: Row #1: Selling rate for item _Test Item is lower than its last purchase rate.
                                        Selling net rate should be atleast 3000.0.Alternatively,
                                        you can disable selling price validation in Selling Settings to bypass
                                        this validation.

### Issue Details:
frappe.exceptions.ValidationError: Row #1: Selling rate for item _Test Item is lower than its last purchase rate.
                                        Selling net rate should be atleast 3000.0.Alternatively,
                                        you can disable selling price validation in Selling Settings to bypass
                                        this validation.

### Traceback:
``` 
FAIL: test_pulling_advance_based_on_debit_to (erpnext.accounts.doctype.sales_invoice.test_sales_invoice.TestSalesInvoice)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/frappe/test-postgres/test-bench/apps/erpnext/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py", line 3883, in test_pulling_advance_based_on_debit_to
    self.assertEqual(1, len(advances))
AssertionError: 1 != 0
```

--- a/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
+++ b/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
@@ -2440,6 +2440,7 @@ class TestSalesInvoice(FrappeTestCase):
 			self.assertEqual(expected_account_values[0], gle.debit)
 			self.assertEqual(expected_account_values[1], gle.credit)
 
+	@if_app_installed("assets")
 	def test_rounding_adjustment_3(self):
 		from erpnext.accounts.doctype.accounting_dimension.test_accounting_dimension import (
 			create_dimension,
@@ -3389,6 +3390,7 @@ class TestSalesInvoice(FrappeTestCase):
 			invoice.reload()
 			self.assertEqual(invoice.status, "Overdue and Discounted")
 
+	@if_app_installed("sales_commission")
 	def test_sales_commission(self):
 		si = frappe.copy_doc(test_records[2])
 
@@ -3856,6 +3858,7 @@ class TestSalesInvoice(FrappeTestCase):
 		check_gl_entries(self, pe.name, expected_gle, nowdate(), voucher_type="Payment Entry")
 		set_advance_flag(company="_Test Company", flag=0, default_account="")
 
+	@if_app_installed("india_compliance")
 	def test_pulling_advance_based_on_debit_to(self):
 		from erpnext.accounts.doctype.payment_entry.test_payment_entry import create_payment_entry
 

--- a/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
+++ b/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
@@ -4879,6 +4879,62 @@ class TestSalesInvoice(FrappeTestCase):
 		self.assertEqual(delivery_note.sales_invoice, sales_invoice.name)
 	
 	def test_sales_invoice_with_update_stock_and_SR_TC_S_027(self):
+		# Set up accounts if they don't exist
+		if not frappe.db.exists("Account", "Stock In Hand - _TC"):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": "Stock In Hand",
+				"parent_account": "Current Assets - _TC",
+				"company": "_Test Company",
+				"account_type": "Stock",
+				"is_group": 0
+			}).insert()
+		
+		if not frappe.db.exists("Account", "Cost of Goods Sold - _TC"):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": "Cost of Goods Sold",
+				"parent_account": "Cost of Goods Sold - _TC",
+				"company": "_Test Company",
+				"account_type": "Cost of Goods Sold",
+				"is_group": 0
+			}).insert()
+		
+		if not frappe.db.exists("Account", "Shipping Charges - _TC"):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": "Shipping Charges",
+				"parent_account": "Indirect Expenses - _TC",
+				"company": "_Test Company",
+				"account_type": "Expense Account",
+				"is_group": 0
+			}).insert()
+		
+		# Set default accounts for test company
+		frappe.db.set_value("Company", "_Test Company", {
+			"default_receivable_account": "Debtors - _TC",
+			"default_income_account": "Sales - _TC",
+			"default_expense_account": "Cost of Goods Sold - _TC",
+			"default_inventory_account": "Stock In Hand - _TC"
+		})
+		
+		# Set up shipping rule account
+		if not frappe.db.exists("Shipping Rule", "_Test Shipping Rule"):
+			shipping_rule = frappe.get_doc({
+				"doctype": "Shipping Rule",
+				"shipping_rule_name": "_Test Shipping Rule",
+				"conditions": [{
+					"doctype": "Shipping Rule Condition",
+					"from_value": 0,
+					"to_value": 100000,
+					"shipping_amount": 200
+				}],
+				"account": "Shipping Charges - _TC"
+			}).insert()
+		else:
+			frappe.db.set_value("Shipping Rule", "_Test Shipping Rule", "account", "Shipping Charges - _TC")
+
+		frappe.db.set_value("Company", "_Test Company", "enable_perpetual_inventory", 1)
 		make_stock_entry(item="_Test Item Home Desktop 100", target="Stores - _TC", qty=10, rate=4000)
 
 		sales_invoice = create_sales_invoice(
@@ -4921,7 +4977,6 @@ class TestSalesInvoice(FrappeTestCase):
 		self.assertEqual(sum([entry.actual_qty for entry in sle]), -4)  
 
 	def test_sales_invoice_with_SR_and_CRN_TC_S_038(self):
-
 		from erpnext.stock.doctype.stock_entry.test_stock_entry import make_stock_entry
 		from erpnext.selling.doctype.sales_order.sales_order import make_delivery_note
 		from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice
@@ -4929,8 +4984,27 @@ class TestSalesInvoice(FrappeTestCase):
 		from erpnext.stock.doctype.delivery_note.test_delivery_note import create_delivery_note
 		from erpnext.accounts.doctype.sales_invoice.test_sales_invoice import create_sales_invoice
 
-		frappe.db.set_value("Company", "_Test Company","enable_perpetual_inventory", 1)
-
+		# Enable perpetual inventory and set up accounts
+		frappe.db.set_value("Company", "_Test Company", "enable_perpetual_inventory", 1)
+		
+		# Create or get inventory account if not exists
+		if not frappe.db.exists("Account", "Stock In Hand - _TC"):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": "Stock In Hand",
+				"parent_account": "Current Assets - _TC",
+				"company": "_Test Company",
+				"account_type": "Stock",
+				"is_group": 0
+			}).insert()
+		
+		# Set default accounts for test company
+		frappe.db.set_value("Company", "_Test Company", {
+			"default_receivable_account": "Debtors - _TC",
+			"default_income_account": "Sales - _TC",
+			"default_expense_account": "Cost of Goods Sold - _TC",
+			"default_inventory_account": "Stock In Hand - _TC"
+		})
 
 		make_stock_entry(item="_Test Item Home Desktop 100", target="Stores - _TC", qty=10, rate=1000)
 
@@ -4965,7 +5039,7 @@ class TestSalesInvoice(FrappeTestCase):
 			item="_Test Item Home Desktop 100",
 			qty=-5,
 			warehouse="Stores - _TC",
-			do_not_save =True
+			do_not_save=True
 		)
 		for i in dn_return.items:
 			i.against_sales_order = sales_order.name
@@ -5001,7 +5075,7 @@ class TestSalesInvoice(FrappeTestCase):
 			rate=3000,
 			is_return=1,
 			return_against=sales_invoice.name,
-			do_not_save =True
+			do_not_save=True
 		)
 		for i in crn.items:
 			i.sales_order = sales_order.name
@@ -5023,7 +5097,34 @@ class TestSalesInvoice(FrappeTestCase):
 		from erpnext.stock.doctype.delivery_note.test_delivery_note import create_delivery_note
 		from erpnext.accounts.doctype.sales_invoice.sales_invoice import make_delivery_note
 
-		frappe.db.set_value("Company", "_Test Company","enable_perpetual_inventory", 1)
+		# Set up required accounts if they don't exist
+		required_accounts = [
+			("Stock In Hand - _TC", "Current Assets - _TC", "Stock"),
+			("Cost of Goods Sold - _TC", "Cost of Goods Sold - _TC", "Cost of Goods Sold"),
+			("Sales - _TC", "Income - _TC", "Income Account"),
+			("Debtors - _TC", "Current Assets - _TC", "Receivable")
+		]
+
+		for account_name, parent_account, account_type in required_accounts:
+			if not frappe.db.exists("Account", account_name):
+				frappe.get_doc({
+					"doctype": "Account",
+					"account_name": account_name.split(" - ")[0],
+					"parent_account": parent_account,
+					"company": "_Test Company",
+					"account_type": account_type,
+					"is_group": 0
+				}).insert()
+
+		# Set default accounts for test company
+		frappe.db.set_value("Company", "_Test Company", {
+			"default_receivable_account": "Debtors - _TC",
+			"default_income_account": "Sales - _TC",
+			"default_expense_account": "Cost of Goods Sold - _TC",
+			"default_inventory_account": "Stock In Hand - _TC"
+		})
+
+		frappe.db.set_value("Company", "_Test Company", "enable_perpetual_inventory", 1)
 		make_stock_entry(item="_Test Item Home Desktop 100", target="Stores - _TC", qty=10, rate=2500)
 
 		sales_invoice = create_sales_invoice(
@@ -5031,8 +5132,8 @@ class TestSalesInvoice(FrappeTestCase):
 			item_code="_Test Item Home Desktop 100",
 			qty=5,
 			rate=3000,
-			income_account = "Sales - _TC",
-			expense_account ="Cost of Goods Sold - _TC",
+			income_account="Sales - _TC",
+			expense_account="Cost of Goods Sold - _TC",
 			update_stock=1
 		)
 

--- a/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
+++ b/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
@@ -6136,6 +6136,7 @@ class TestSalesInvoice(FrappeTestCase):
 				"due_date": today(),
 				"currency": "INR",
 				"selling_price_list": price_list,
+				"taxes_and_charges" :  "Output GST In-state - TC-1",
 				"items": [
 					{
 						"item_code": item_name,
@@ -6168,6 +6169,7 @@ class TestSalesInvoice(FrappeTestCase):
 
 		pi = make_inter_company_purchase_invoice(si.name)
 		pi.bill_no = "test bill"
+		pi.taxes_and_charges =  "Input GST In-state - TC-3"
 		pi.insert()
 		pi.submit()
 		self.assertEqual(pi.company, child_company)
@@ -6326,6 +6328,7 @@ class TestSalesInvoice(FrappeTestCase):
 				"transaction_date": today(),
 				"set_warehouse": "Stores - TC-1",
 				"selling_price_list": price_list,
+				"taxes_and_charges":"Output GST In-state - TC-1",
 				"items": [
 					{
 						"item_code": item.item_code,
@@ -6346,6 +6349,7 @@ class TestSalesInvoice(FrappeTestCase):
 		po = make_inter_company_purchase_order(so.name)
 		po.schedule_date = today()
 		po.set_warehouse = "Stores - TC-3"
+		po.taxes_and_charges =  "Input GST In-state - TC-3"
 
 		po.insert()
 		po.submit()
@@ -6391,6 +6395,7 @@ class TestSalesInvoice(FrappeTestCase):
 
 
 		pr = make_inter_company_purchase_receipt(dn.name)
+		pr.taxes_and_charges =  "Input GST In-state - TC-3"
 		pr.insert()
 		pr.submit()
 
@@ -6444,6 +6449,7 @@ class TestSalesInvoice(FrappeTestCase):
 
 		pi = make_inter_company_purchase_invoice(si.name)
 		pi.bill_no = "test bill"
+		pi.taxes_and_charges =  "Input GST In-state - TC-3"
 		pi.insert()
 		pi.submit()
 
@@ -7316,6 +7322,8 @@ def create_company_and_supplier():
 			}
 		).insert()
 
+	create_test_tax_data()
+
 	return {
 		"parent_company": parent_company,
 		"child_company": child_company,
@@ -7323,6 +7331,104 @@ def create_company_and_supplier():
 		"customer": customer,
 		"price_list": price_list
 	}
+
+def create_test_tax_data():
+	company = "Test Company-1122"
+	company_abbr = "TC-1" 
+	child_company = "Test Company-3344"
+	child_company_abbr="TC-3"
+
+	required_accounts_parent= [
+        ("Output Tax SGST", "Duties and Taxes"),
+        ("Output Tax CGST", "Duties and Taxes")
+    ]
+
+	required_accounts_child= [
+        ("Input Tax SGST", "Duties and Taxes"), 
+        ("Input Tax CGST", "Duties and Taxes")
+    ]
+    
+	if not frappe.db.exists("Account", f"Duties and Taxes - {company_abbr}"):
+		frappe.get_doc({
+			"doctype": "Account",
+			"account_name": "Duties and Taxes",
+			"parent_account": "Indirect Expenses - " + company_abbr,
+			"company": company,
+			"is_group": 1
+		}).insert()
+	
+	if not frappe.db.exists("Account", f"Duties and Taxes - {child_company_abbr}"):
+		frappe.get_doc({
+			"doctype": "Account",
+			"account_name": "Duties and Taxes",
+			"parent_account": "Indirect Expenses - " + child_company_abbr,
+			"company": child_company,
+			"is_group": 1
+		}).insert()
+
+	for account_name, parent in required_accounts_parent:
+		full_name = f"{account_name} - {company_abbr}"
+		if not frappe.db.exists("Account", full_name):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": account_name,
+				"parent_account": f"{parent} - {company_abbr}",
+				"company": company,
+				"account_type": "Tax"
+			}).insert()
+	
+	for account_name, parent in required_accounts_child:
+		full_name = f"{account_name} - {child_company_abbr}"
+		if not frappe.db.exists("Account", full_name):
+			frappe.get_doc({
+				"doctype": "Account",
+				"account_name": account_name,
+				"parent_account": f"{parent} - {child_company_abbr}",
+				"company": child_company,
+				"account_type": "Tax"
+			}).insert()
+
+	if not frappe.db.exists("Sales Taxes and Charges Template", "Output GST In-state - TC-1"):
+		frappe.get_doc({
+			"doctype": "Sales Taxes and Charges Template",
+			"title": "Output GST In-state",
+			"company": company,
+			"taxes": [
+				{
+					"charge_type": "On Net Total",
+					"account_head": f"Output Tax SGST - {company_abbr}",
+					"rate": 9,
+					"description": f"SGST - {company_abbr}"
+				},
+				{
+					"charge_type": "On Net Total", 
+					"account_head": f"Output Tax CGST - {company_abbr}",
+					"rate": 9,
+					"description": f"CGST - {company_abbr}"
+				}
+			]
+		}).insert()
+
+	if not frappe.db.exists("Purchase Taxes and Charges Template", "Input GST In-state - TC-1"):
+		frappe.get_doc({
+			"doctype": "Purchase Taxes and Charges Template",
+			"title": "Input GST In-state",
+			"company": child_company,
+			"taxes": [
+				{
+					"charge_type": "On Net Total",
+					"account_head": f"Input Tax CGST - {child_company_abbr}",
+					"rate": 9,
+					"description": f"CGST - {child_company_abbr}"
+				},
+				{
+					"charge_type": "On Net Total",
+					"account_head": f"Input Tax SGST - {child_company_abbr}",
+					"rate": 9,
+					"description": f"SGST - {child_company_abbr}"
+				}
+			]
+		}).insert()
 
 def get_active_fiscal_year():
 	from datetime import datetime

--- a/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
+++ b/erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py
@@ -7,7 +7,7 @@ import json
 import frappe
 from frappe import qb
 from frappe.model.dynamic_links import get_dynamic_link_map
-from frappe.tests.utils import FrappeTestCase, change_settings
+from frappe.tests.utils import FrappeTestCase, change_settings, if_app_installed
 from frappe.utils import add_days, flt, format_date, getdate, nowdate, today
 from erpnext.stock.get_item_details import get_item_tax_map
 import erpnext
@@ -4682,6 +4682,7 @@ class TestSalesInvoice(FrappeTestCase):
 				voucher_type="Payment Entry",
 				posting_date=pe.posting_date	
 			)
+	@if_app_installed("india_compliance")
 	def test_sales_invoice_without_sales_order_with_gst_TC_S_016(self):
 		from erpnext.stock.doctype.warehouse.test_warehouse import create_warehouse
 
@@ -4757,6 +4758,7 @@ class TestSalesInvoice(FrappeTestCase):
 				dn_acc_debit = frappe.db.get_value('GL Entry', {'voucher_type': 'Delivery Note', 'voucher_no': dn.name, 'account': 'Cost of Goods Sold - _TIRC'}, 'debit')
 				self.assertEqual(dn_acc_debit, qty_change[0].get("valuation_rate") * 4)
 	
+	@if_app_installed("india_compliance")
 	def test_sales_invoice_with_update_stock_checked_with_gst_TC_S_017(self): 
 		from erpnext.stock.doctype.warehouse.test_warehouse import create_warehouse
 
@@ -5811,6 +5813,7 @@ class TestSalesInvoice(FrappeTestCase):
 		si = make_sales_invoice(so.name)
 		self.assertEquals(si.payment_terms_template,"_Test Payment Term Template")
 	
+	@if_app_installed("india_compliance")
 	def test_generate_sales_invoice_with_items_different_gst_rates_TC_ACC_131(self):
 		from erpnext.accounts.doctype.payment_entry.test_payment_entry import make_test_item
 		import json

--- a/erpnext/maintenance/doctype/maintenance_schedule/test_maintenance_schedule.py
+++ b/erpnext/maintenance/doctype/maintenance_schedule/test_maintenance_schedule.py
@@ -6,6 +6,7 @@ import unittest
 import frappe
 from frappe.utils import format_date
 from frappe.utils.data import add_days, formatdate, today
+from frappe.tests.utils import if_app_installed
 
 from erpnext.maintenance.doctype.maintenance_schedule.maintenance_schedule import (
 	get_serial_nos_from_schedule,
@@ -123,6 +124,7 @@ class TestMaintenanceSchedule(unittest.TestCase):
 
 		frappe.db.rollback()
 
+	@if_app_installed("sales_commission")
 	def test_schedule_with_serials(self):
 		# Checks whether serials are automatically updated when changing in items table.
 		# Also checks if other fields trigger generate schdeule if changed in items table.

--- a/erpnext/manufacturing/doctype/blanket_order/test_blanket_order.py
+++ b/erpnext/manufacturing/doctype/blanket_order/test_blanket_order.py
@@ -255,7 +255,9 @@ class TestBlanketOrder(FrappeTestCase):
 	def test_blanket_order_to_sales_invoice_TC_S_054(self):
 		from erpnext.selling.doctype.sales_order.sales_order import make_delivery_note
 		from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice
+		from erpnext.stock.doctype.stock_entry.stock_entry_utils import make_stock_entry
 		frappe.flags.args.doctype = "Sales Order"
+		make_stock_entry(item_code="_Test Item", qty=100, rate=5000, target="_Test Warehouse - _TC")
 
 		bo = make_blanket_order(blanket_order_type="Selling",quantity=50,rate=1000)
 		so = make_order(bo.name)

--- a/erpnext/selling/doctype/sales_order/test_sales_order.py
+++ b/erpnext/selling/doctype/sales_order/test_sales_order.py
@@ -3313,17 +3313,7 @@ class TestSalesOrder(AccountsTestMixin, FrappeTestCase):
 		from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice
 		from erpnext.stock.doctype.quality_inspection.test_quality_inspection import create_quality_inspection
 
-		item = frappe.get_doc({
-			"doctype": "Item",
-			"item_code": "_Test Item for quality inspection",
-			"item_name": "_Test Item for quality inspection",
-			"stock_uom": "Nos",
-			"is_stock_item": 1,
-			"opening_stock": 1000,
-			"valuation_rate": 100,
-			"inspection_required_before_delivery": 1,
-			"gst_hsn_code": "90183100"
-		}).insert()
+		item=make_item("_Test Item for quality inspection", {"is_stock_item": 1,"inspection_required_before_delivery":1})
 
 		parameters = ["_Test Needle Shape", "_Test Syringe Shape", "_Test Plastic Clarity", "_Test Syringe Length"]
 		for param in parameters:

--- a/erpnext/stock/doctype/delivery_note/test_delivery_note.py
+++ b/erpnext/stock/doctype/delivery_note/test_delivery_note.py
@@ -2700,7 +2700,7 @@ class TestDeliveryNote(FrappeTestCase):
 	def test_dn_submission_TC_SCK_148(self):
 		from erpnext.stock.doctype.warehouse.test_warehouse import create_warehouse
 		from erpnext.buying.doctype.supplier.test_supplier import create_supplier
-    from erpnext.selling.doctype.sales_order.test_sales_order import get_or_create_fiscal_year
+		from erpnext.selling.doctype.sales_order.test_sales_order import get_or_create_fiscal_year
 		# from erpnext_crm.erpnext_crm.doctype.lead.lead import make_customer
 		from erpnext.stock.doctype.delivery_note.test_delivery_note import create_delivery_note
 		"""Test Purchase Receipt Creation, Submission, and Stock Ledger Update"""

--- a/erpnext/stock/doctype/delivery_note/test_delivery_note.py
+++ b/erpnext/stock/doctype/delivery_note/test_delivery_note.py
@@ -2691,7 +2691,7 @@ class TestDeliveryNote(FrappeTestCase):
 			company="_Test Company",
 		)
 		create_warehouse(
-			warehouse_name=" _Test Warehouse - _TC",
+			warehouse_name="_Test Warehouse - _TC",
 			properties={"parent_warehouse": "All Warehouses - _TC"},
 			company="_Test Company",
 		)

